File: 2019/fsharp/Day01.fs
namespace AdventOfCode.FSharp.Y2019

// The Tyranny of the Rocket Equation
module Day01 =
    open System
    open AdventOfCode.FSharp.Util

    let part1 inputs =
        let f x = max ((x / 3) - 2) 0
        let q = inputs |> Seq.map f
        q |> Seq.sum

    let part2 inputs =
        let f x = max ((x / 3) - 2) 0

        let g x =
            let y = f x
            if y > 0 then Some(y, y) else None

        let h x = 
            List.unfold g x 
            |> List.sum

        inputs |> Seq.map h |> Seq.sum

    let run (input: byte array) output =
        let nums = parseInts input
        part1 nums |> string |> output 1
        part2 nums |> string |> output 2

File: 2019/csharp/Day01.cs
namespace AdventOfCode.CSharp.Y2019;

public static class Day01
{
    public static void Run(byte[] input, Action<int, string> output)
    {
        var Input = Encoding.UTF8
            .GetString(input)
            .Split('\n', StringSplitOptions.RemoveEmptyEntries);

        long Part1()
        {
            return Input.Select(long.Parse).Select(x => Math.Max((x / 3) - 2, 0)).Sum();
        }

        int Part2()
        {
            Func<int, int> f = x => Math.Max((x / 3) - 2, 0);
            Func<int, int> g = x =>
            {
                var s = 0;
                while (x > 0)
                {
                    x = f(x);
                    s += x;
                }
                return s;
            };

            return Input.Select(int.Parse).Select(g).Sum();
        }

        output(1, Part1().ToString());
        output(2, Part2().ToString());
    }
}